robot_smach_states/src/robot_smach_states/navigation/find.py

```diff
@@ -0,0 +1,123 @@
+__author__ = 'rokus'
+
+import smach, rospy
+import robot_smach_states as states
+from robot_smach_states.util.designators import VariableDesignator, EdEntityDesignator
+from robot_skills.classification_result import ClassificationResult
+from robocup_knowledge import load_knowledge
+
+def entities_from_description(robot, entity_description, list_of_entity_ids=None ):
+    '''
+    Query entities and return those that satisfy the given description
+
+    @param robot: The robot object
+    @param entity_descr: A dict that contains a 'type' field
+    @param list_of_entity_ids: A list of entity ids to choose from (for example a result of a segment)
+
+    @return: entities
+        entities  - list of entities that fulfill the description
+                    (each element has type Entity)
+    '''
+    knowledge = load_knowledge('common')
+
+    if not isinstance(list_of_entity_ids, list):
+        return []
+
+    if not isinstance(entity_description, dict):
+        return []
+    if 'type' not in entity_description:
+        return []
+
+    # Get all entities from the world model
+    entities = robot.ed.get_entities()
+
+    # TODO: hack because ed maintains all labels that were ever assigned to an entity in the .types field
+    if entity_description['type'] == 'person':
+        entities = [e for e in entities if e.is_a('possible_human')]
+    else:
+        # Select entities based on the description (only type for now)
+        try:
+            if entity_description['type'] in knowledge.object_categories:
+                entities = [e for e in entities if knowledge.object_category(e.type) == entity_description['type']]
+            else:
+                entities = [e for e in entities if e.type == entity_description['type']]
+        except:
+            entities = [e for e in entities if e.type == entity_description['type']]
+
+
+        # If we have a list of entities to choose from, select based on that list
+        if list_of_entity_ids:
+            entities = [e for e in entities if e.id in list_of_entity_ids]
+
+    # Sort entities by distance
+    robot_pos = robot.base.get_location().frame.p
+    entities = sorted(entities, key=lambda entity: entity.distance_to_2d(robot_pos))
+
+    return entities
+
+
+class CheckIfDescribedEntityAvailable(smach.State):
+    """
+    Check if a described entity is in the world model
+    """
+    def __init__(self,
+                 robot,
+                 description_designator,
+                 found_entity_designator,
+                 candidate_entities_designator=None):
+        smach.State.__init__(self, outcomes=['succeeded', 'failed'])
+        self._robot = robot
+        self._description_designator = description_designator
+        self._found_entity_designator = found_entity_designator
```

Comment:
You may want to check that the designator is really writeable. If you need a VariableDesignator at all and not e.g. a to be defined EntityFromDescriptionDesignator